Make SchemaCache loading faster

Edouard-chin opened the pull request:
Make SchemaCache loading faster:

- A benchmark of the time it takes to load the schema cache will give
  you a first overview of what this PR is about:

  ```
  Warming up --------------------------------------
               Before     1.000  i/100ms
               Now        1.000  i/100ms
  Calculating -------------------------------------
               Before     0.872  (± 0.0%) i/s -      9.000  in  10.381223s
               Now        7.209  (±13.9%) i/s -     70.000  in  10.067905s
  ```
  This benchmark was done on a SchemaCache that contained 428 tables
  and more than 5k columns.

  I first dug into SchemaCache in our application has we were dumping
  the Schema and loading it manually. It felt weird and I decided to
  get rid of that code in order to dump it like like Rails does,
  with Psych.
  By curiosity I benchmarked and realized that our implementation was
  8 times faster.

  I'm not familiar enough with Psych to understand why it was so much
  slower, but after some exploration in a stackprof dump it seemed
  obvious that the times was majoritaly spent allocating objects and
  parsing node/tree.

  This implementation serializes the SchemaCache with the raw data we
  get from the 'SHOW FULL FIELDS' query.
  A new  SchemaCacheSerializer then iterate over the YAML and recreate the Column
  and IndexDefinition object one by one.

  The dumped schema cache YAML will look like this
<details>
<summary> Please expand </summary>

```yml
:columns:
  achievements:
  - :Field: id
    :Type: bigint(20)
    :Collation: 
    :Null: 'NO'
    :Key: PRI
    :Default: 
    :Extra: auto_increment
    :Privileges: select,insert,update,references
    :Comment: ''
:indexes:
  achievements:
  - :Table: achievements
    :Non_unique: 0
    :Key_name: PRIMARY
    :Seq_in_index: 1
    :Column_name: id
    :Collation: A
    :Cardinality: 0
    :Sub_part: 
    :Packed: 
    :Null: ''
    :Index_type: BTREE
    :Comment: ''
    :Index_comment: ''
:data_sources:
  achievements: true
:primary_keys:
  achievements: id
:version: 201901221501932  
```
</details>

  The SchemaCacheSerializer also has an option to skip tables that you
  don't want to include in the dump. It's useful if for instance you
  have tables that are related to schema change (like the internal
  ar_internal_metadata, but also others if you use LHMs).

  The Serializer support SchemaCache dumped with Psych, however it's
  deprecated and should be remove in the next version of Rails.

  In order to achieve this I had to extract code related to building indexes into their own methods.
  It created big chunks that I grouped into a commit to make things easier for review.


@rafaelfranca @kaspth  cc/ @byroot (I was thinking about adding the coder plug in a latter PR, but maybe I should do it here directly?)

Edouard-chin committed:
Extract indexes related code into their own method:

- Following the pattern to grab the columns definitions (one method to
  to do the `SHOW FULL FIELDS` query and another to build the column
  object from the result).

  Making this change in order to have an easier way to dump the
  indexes without creating IndexDefinition object as well as
  rebuilding them afterwards without making a query.

Edouard-chin committed:
Introduce the SchemaCacheSerializer:

- A benchmark of the time it takes to load the schema cache will give
  you a first overview of what this commit is about:

  ```
  Warming up --------------------------------------
               Before     1.000  i/100ms
               Now        1.000  i/100ms
  Calculating -------------------------------------
               Before     0.872  (± 0.0%) i/s -      9.000  in  10.381223s
               Now        7.209  (±13.9%) i/s -     70.000  in  10.067905s
  ```
  This benchmark was done on a SchemaCache that contained 428 tables
  and more than 5k columns.

  I first dug into SchemaCache in our application has we were dumping
  the Schema and loading it manually. It felt weird and I decided to
  get rid of that code in order to dump it like like Rails does,
  with Psych.
  By curiosity I benchmarked and realized that our implementation was
  8 times faster.

  I'm not familiar enough with Psych to understand why it was so much
  slower, but after some exploration in a stackprof dump it seemed
  obvious that the times was majoritaly spent allocating objects and
  parsing node/tree.

  This implementation serializes the SchemaCache with the raw data we
  get from the 'SHOW FULL FIELDS' query.
  The serializer then iterate over the YAML and recreate the Column
  and IndexDefinition object one by one.

  The SchemaCacheSerializer also has an option to skip tables that you
  don't want to include in the dump. It's useful if for instance you
  have tables that are related to schema change (like the internal
  ar_internal_metadata, but also others if you use LHMs).

  The Serializer support SchemaCache dumped with Psych, however it's
  deprecated and should be remove in the next version of Rails.

Edouard-chin committed:
Add a new `load_schema_cache` method:

- Having a separate method in the DatabaseTasks to load the schema
  make things easier for application that have sharded databases and
  need to load the schema cache for each shard.

  This commit also make use of the SchemaCacheSerializer to dump the
  schema cache.

rafaelfranca commented on activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:
```diff
@@ -1198,6 +1198,10 @@ def create_schema_dumper(options) # :nodoc:
         SchemaDumper.create(self, options)
       end
 
+      def schema_cache_serializer
```
`# :nodoc:`?

rafaelfranca commented on activerecord/lib/active_record/schema_cache_serializer.rb:
```diff
@@ -0,0 +1,75 @@
+        indexes = {}
+
+        serialized_schema.each do |table, serialized_indexes|
+          indexes[table] = @connection.send(:new_indexes_from_fields, serialized_indexes, table)
```
`new_indexes_from_fields` should be public but with `:nodoc:`

rafaelfranca commented on activerecord/lib/active_record/railtie.rb:
```diff
@@ -31,6 +31,8 @@ class Railtie < Rails::Railtie # :nodoc:
     config.active_record.sqlite3 = ActiveSupport::OrderedOptions.new
     config.active_record.sqlite3.represent_boolean_as_integer = nil
 
+    config.active_record.schema_cache_tables_to_skip = []
```
`table_to_ignire_on_the_schema_cache`

rafaelfranca commented on activerecord/lib/active_record/schema_cache_serializer.rb:
```diff
@@ -0,0 +1,75 @@
+        next if tables_to_skip.include?(table)
+
+        serialized_schema[:columns][table] = @connection.send(:column_definitions, table).to_a
+        serialized_schema[:indexes][table] = @connection.send(:index_definitions, table).to_a
```
public but with nodoc.

rafaelfranca commented on activerecord/lib/active_record/schema_cache_serializer.rb:
```diff
@@ -0,0 +1,75 @@
+
+        serialized_schema.each do |table, serialized_columns|
+          columns[table] = serialized_columns.map do |sc|
+            @connection.send(:new_column_from_field, table, sc)
```
same with the send, if we need to use internally it should be public with nodoc.

matthewd commented:
Separating the queries from the code that uses them seems like a pain for maintenance (and just reading). Caching the intermediate value seems likely to make future upgrades & changes that touch those queries more difficult to deal with. 

If we're going to use a Rails-version-fragile format, why not switch back to Marshal? Cross-version compatibility is exactly why Shopify introduced the YAML format in #27042, and AFAICS this change takes us straight back to "schema load will probably explode in some nonspecific way if you load a dump from a different Rails version". We seem to be ending up back where we started wrt compatibility (and performance), but at +1000 lines of code. :confused:

rafaelfranca commented:
> "schema load will probably explode in some nonspecific way if you load a dump from a different Rails version"

That actually don't happen with this approach. It happens with marshal because it depends on the internal structure of the objects dumped, if you change the object you change the result of the dump. This approach, like YAML dump allow us to control the format of the dump, which we can maintain always compatible.

I really don't see how this is a "Rails-version-fragile format", is is basically the current YAML, but without all the overhead of the object building present on the plain `YAML.dump(schema_cache_object)`

rafaelfranca commented:
What is strange is that manually allocating the same objects is less expensive than allocating those objects using psych. There is something off here.

matthewd commented:
But (unless I'm misreading?) it's dumping the raw results of the schema queries -- which regularly change between versions.

It seems easier to use a custom Marshal dump format on the objects instead of grabbing internal data mid-flight.

Edouard-chin commented:
Thanks for the quick review Rafael, Matthew! 

> But (unless I'm misreading?) it's dumping the raw results of the schema queries -- which regularly change between versions.

When you say it change between versions, we are talking about MySQL/Postgres... versions right?
I wasn't expecting the schema result to change much between versions 🤔

matthewd commented:
No, I mean it (the query, and therefore the resulting data) changes between Rails versions. e.g. these lines: https://github.com/rails/rails/blob/da2c92377c7f36469e56dc25a4ac0604157ac778/activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb#L88-L100

from a glance in blame, substantively changed in at least 5.2 (#19090) and 5.0 (#22911) -- and that's just looking at the top-most commits from those lines, without checking history at all.

>  it seems like it's mostly spending time parsing the YAML

This is a great point. And having just actually looked at a real-world schema dump file, I can rather see why it might not be happy. 😵

I'd like to revise my previous suggestion, such that we instead write a custom YAML dumper (instead of a custom Marshal dumper). I think it will be clearer to explicitly define, inside e.g. the Column classes, what they need dumped and how to reconstruct them, instead of intercepting the internal query result. That also allows smooth version compatibility, because it's behaviour separated from how the data is originally obtained. (And I'm obviously working on the premise that a consciously-constructed dump format will be far more compact, and thus avoid the performance issues with the current "all the things"-looking raw dump.)

casperisfine commented:
> such that we instead write a custom YAML dumper 

Something we haven't really mentioned yet, but that we'd like to submit soon is to allow to customize the encoder.

Since we have a lot of shards, we're loading quite a lot of schema caches and would like to use something else than YAML for serialization and deserialization (we're thinking messagepack).

So if the YAML only contained basic ruby types, it would be much simpler for us.

That being said, if it makes everything more complicated then we understand.

Edouard-chin commented:
Thanks for the suggestion @matthewd .

I actually tried what you described in my first implementation, but has to resign because I hit a wall (also the amount of LOC to implement this was way higher since each adapter needed their own serializer to define what field they need to dump and how to reconstruct them).

The main issue with this approach was about serializing/reconstructing IndexDefinition (ID) object. For instance in MySQL the logic to create IndexDefinition object is non trivial and one field in the Raw result doesn't correspond to an attribute in the ID object, thus I basically need to "reverse engineer" [this code](https://github.com/rails/rails/blob/e9e976fbc4c35e5800a10d26f01aaa9c70227de7/activerecord/lib/active_record/connection_adapters/mysql/schema_statements.rb#L12-L49) to be able to reconstruct ID from stratch 

By doing this I felt like the SchemaCache serializer had to be aware of the internal of how ID are create in the first place and was tighten to it's implementation

kamipo commented on activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb:
```diff
@@ -6,50 +6,7 @@ module SQLite3
-              orders: orders
-            )
-          end.compact
+          new_indexes_from_fields(index_definitions(table_name), table_name)
```
Why don't you put `new_indexes_from_fields` just below `indexes`?
It reduces the diff dramatically.

```diff
diff --git a/activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb b/activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb
index e64e995e1a..a8ff4f01db 100644
--- a/activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb
+++ b/activerecord/lib/active_record/connection_adapters/sqlite3/schema_statements.rb
@@ -6,7 +6,11 @@ module SQLite3
       module SchemaStatements # :nodoc:
         # Returns an array of indexes for the given table.
         def indexes(table_name)
-          exec_query("PRAGMA index_list(#{quote_table_name(table_name)})", "SCHEMA").map do |row|
+          new_indexes_from_fields(index_definitions(table_name), table_name)
+        end
+
+        def new_indexes_from_fields(fields, table_name)
+          fields.map do |row|
             # Indexes SQLite creates implicitly for internal use start with "sqlite_".
             # See https://www.sqlite.org/fileformat2.html#intschema
             next if row["name"].starts_with?("sqlite_")
```

Edouard-chin replied:
👍 I mainly wanted to add this next to the `def new_column_from_field` method. Happy to move it up though !

Edouard-chin committed:
Make few internal methods public and mark them as nodoc:

- Decided to use the public statement to avoid creating too many large git chunk
- Also rename `schema_cache_tables_to_skip` -> `table_to_ignore_on_schema_cache`

rafaelfranca commented:
@Edouard-chin is this still worth? If so, what is missing?
